src/object.hpp

```diff
@@ -178,20 +191,323 @@ template <typename Derived> class readable_object {
+               obj.intValue <= limits::max();
+    }
+
+    // Convert the underlying type to the requested type, converters are defined
+    // in the object_converter header
+    template <typename T> T convert() const;
+
+    [[nodiscard]] owned_object clone() const;
+
 private:
     readable_object() = default;
 
     friend Derived;
 };
 
-// TODO move object view and object converter header
+// Temporary abstraction, this will be removed once the keys and values are
+// split within ddwaf_object.
+class object_key {
+public:
+    // The default constructor results in a key without value
+    object_key() = default;
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_key(detail::object *underlying_object) : obj_(underlying_object) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_key(const detail::object *underlying_object) : obj_(underlying_object) {}
+
+    ~object_key() = default;
+    object_key(const object_key &) = default;
+    object_key(object_key &&) = default;
+    object_key &operator=(const object_key &) = default;
+    object_key &operator=(object_key &&) = default;
+
+    [[nodiscard]] const char *data() const noexcept
+    {
+        return obj_ != nullptr ? obj_->parameterName : nullptr;
+    }
+
+    [[nodiscard]] std::size_t size() const noexcept
+    {
+        if (obj_ == nullptr || obj_->parameterName == nullptr) {
+            [[unlikely]] return 0;
+        }
+        return static_cast<std::size_t>(obj_->parameterNameLength);
+    }
+
+    [[nodiscard]] bool empty() const noexcept { return size() == 0; }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, std::string> || std::is_same_v<T, std::string_view>
+    {
+        if (obj_ == nullptr || obj_->parameterName == nullptr) {
+            [[unlikely]] return {};
+        }
+        return {obj_->parameterName, static_cast<std::size_t>(obj_->parameterNameLength)};
+    }
+
+    template <typename T>
+    bool operator==(const T &other) const
+        requires(std::is_same_v<T, std::string_view> || std::is_same_v<T, object_key>)
+    {
+        auto s = size();
+        return s == other.size() && memcmp(data(), other.data(), s) == 0;
+    }
+
+protected:
+    const detail::object *obj_{nullptr};
+};
+
+class object_view final : public readable_object<object_view> {
+public:
+    // The default constructor results in a view without value
+    object_view() = default;
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const detail::object &underlying_object) : obj_(&underlying_object) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const detail::object *underlying_object) : obj_(underlying_object) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const owned_object &ow);
```

Comment:
I think you should delete `object_view(owned_object&&)` , as that's always wrong